File: backend/src/main/java/com/example/backend/model/Event.java
package com.example.backend.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;

import java.time.LocalDate;
import java.util.Date;
@Entity
public class Event {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @NotBlank(message = "Title is required")

    private String title;
    @Size(max = 500, message = "Description is too long")

    private String description;

    @NotNull(message = "Date is required")
    @FutureOrPresent(message = "Date must be today or in the future")
    private LocalDate date;
    @NotBlank(message = "Location is required")

    private String location;
    @Min(value = 1, message = "Max participants must be at least 1")

    private int maxParticipants;
    @NotNull(message = "Organizer must be specified")

    @ManyToOne
    @JoinColumn(name = "organizer_id")
    private User organizer;

    public Event(){}

    public Event(Long id, String title, String description, LocalDate date, String location, int maxParticipants, User organizer) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.date = date;
        this.location = location;
        this.maxParticipants = maxParticipants;
        this.organizer = organizer;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public int getMaxParticipants() {
        return maxParticipants;
    }

    public void setMaxParticipants(int maxParticipants) {
        this.maxParticipants = maxParticipants;
    }

    public User getOrganizer() {
        return organizer;
    }

    public void setOrganizer(User organizer) {
        this.organizer = organizer;
    }
}

File: mobile_app/lib/models/event.dart
import 'package:flutter/material.dart';
import 'user.dart';

class Event {
  final int id;
  final String title;
  final String description;
  final DateTime date;
  final String location;
  final int maxParticipants;
  final User organizer;

  Event({
    required this.id,
    required this.title,
    required this.description,
    required this.date,
    required this.location,
    required this.maxParticipants,
    required this.organizer,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['id'],
      title: json['title'],
      description: json['description'] ?? '',
      date: DateTime.parse(json['date']), // Backend sends YYYY-MM-DD
      location: json['location'],
      maxParticipants: json['maxParticipants'],
      organizer: User.fromJson(json['organizer']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'date': date.toIso8601String().split('T')[0], // Send as YYYY-MM-DD
      'location': location,
      'maxParticipants': maxParticipants,
      'organizer': organizer.toJson(),
    };
  }

  // For creating a new event
  factory Event.create({
    required String title,
    required String description,
    required DateTime date,
    required String location,
    required int maxParticipants,
    required User organizer,
  }) {
    return Event(
      id: 0, // Will be set by backend
      title: title,
      description: description,
      date: date,
      location: location,
      maxParticipants: maxParticipants,
      organizer: organizer,
    );
  }

  Event copyWith({
    int? id,
    String? title,
    String? description,
    DateTime? date,
    String? location,
    int? maxParticipants,
    User? organizer,
  }) {
    return Event(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      date: date ?? this.date,
      location: location ?? this.location,
      maxParticipants: maxParticipants ?? this.maxParticipants,
      organizer: organizer ?? this.organizer,
    );
  }

  // Validation method to match backend constraints
  String? validate() {
    if (title.isEmpty) {
      return 'Title is required';
    }
    if (description.length > 500) {
      return 'Description is too long (max 500 characters)';
    }
    if (date.isBefore(DateTime.now().subtract(Duration(days: 1)))) {
      return 'Date must be today or in the future';
    }
    if (location.isEmpty) {
      return 'Location is required';
    }
    if (maxParticipants < 1) {
      return 'Max participants must be at least 1';
    }
    if (organizer.id == 0) {
      return 'Organizer must be specified';
    }
    return null;
  }
} 